fda/basis.py

```diff
@@ -1268,6 +1376,267 @@ def evaluate(self, eval_points, derivative=0):
+                to all samples.
+            ext (str or int, optional): Controls the extrapolation mode for
+                elements outside the domain range.
+                * If ext=0 or 'default' uses the default method defined in the
+                  basis.
+                * If ext=1 or 'extrapolation' uses the extrapolated values.
+                * If ext=2 or 'periodic' extends the domain range periodically.
+                * If ext=3 or 'const' uses the boundary value.
+                * If ext=4 or 'slice' avoids extrapolation restricting the
+                domain.
+                The default value is 'default'.
+            tfine (array_like, optional): Set of points where the
+                functions are evaluated to obtain the discrete
+                representation of the object to integrate. If an empty list is
+                passed it calls numpy.linspace with bounds equal to the ones defined
+                in fd.domain_range and the number of points the maximum
+                between 201 and 10 times the number of basis plus 1.
+            **kwargs: Keyword arguments to be passed to :meth:`from_data`.
+
+        Returns:
+            :obj:`FDataBasis` with the curves :math:`f_i(t + \delta_i)`.
+        """
+
+        if not len(tfine):  # Grid to discretize the function
+            nfine = max(self.nbasis*10+1, 201)
+            tfine = numpy.linspace(self.basis.domain_range[0],
+                                   self.basis.domain_range[1],
+                                   nfine)
+        else:
+            tfine = numpy.asarray(tfine)
+
+        if numpy.isscalar(shifts): # Special case, al curves have the same shift
+
+            _basis = self.basis.rescale((self.basis.domain_range[0] + shifts,
+                                         self.basis.domain_range[1] + shifts))
+
+            return FDataBasis.from_data(self.evaluate(tfine), tfine + shifts,
+                                        _basis, **kwargs)
+
+        elif len(shifts) != self.nsamples:
+            raise ValueError("shifts vector ({}) must have the same length "
+                             "than the number of samples ({})"
+                             .format(len(shifts), self.nsamples))
+
+        periodic, extrapolation = _check_extrapolation(self, ext)
+
+        if not periodic:
+            a = self.domain_range[0] - min(numpy.min(shifts), 0)
+            b = self.domain_range[1] - max(numpy.max(shifts), 0)
+            domain = (a, b)
+            tfine = tfine[numpy.logical_and(tfine >= a, tfine <= b)]
+        else:
+            domain = self.domain_range
+
+        # Matrix of shifted values
+        _data_matrix = self.evaluate_shifted(tfine, shifts, ext=extrapolation)
+        _basis = self.basis.rescale(domain)
+
+        return FDataBasis.from_data(_data_matrix, tfine, _basis, **kwargs)
+
+    def shift_registration(self, maxiter=5, tol=1e-2, ext="default",
```

Comment:
I am not sure it is a good idea to repeat functions as methods. I think that is more useful to have a few core small methods, and implement the functions using them, but we can discuss it. See https://github.com/numpy/numpy/issues/7452#issue-142974954

Reply:
Changed in e492eda.